staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go

```diff
@@ -40,6 +50,40 @@ func handleError(w http.ResponseWriter, r *http.Request, err error) {
 	glog.Errorf(err.Error())
 }
 
+// requestWatermark is used to trak maximal usage of inflight requests.
+type requestWatermark struct {
+	lock                                 sync.Mutex
+	readOnlyWatermark, mutatingWatermark int
+}
+
+func (w *requestWatermark) record(readOnlyVal, mutatingVal int) {
+	w.lock.Lock()
+	defer w.lock.Unlock()
```

Comment:
this is still a blocking lock. might be a good use for non-locking updates (would want to benchmark with contention)

```
oldVal := atomic.LoadInt64(&w.readOnlyWatermark)
while oldVal < newVal && !atomic.CompareAndSwapInt64(&w.readOnlyWatermark, oldVal, newVal) {
  oldVal = atomic.LoadInt64(&w.readOnlyWatermark)
}
```

Reply:
My feeling is that 3 atomic operations are worse than lock, but I agree benchmark would be needed to be sure, but even that may change in the future versions of go. Which is why I'd want to proceed with the current version. Locking is now done after processing the request and releasing the "inflight" token, which should prevent this impacting the critical path in the apiserver. Does that sound good to you?

Reply:
> My feeling is that 3 atomic operations are worse than lock

that would only happen on requests that bumped the high water mark... the majority would be  a single read.

Reply:
> Which is why I'd want to proceed with the current version. Locking is now done after processing the request and releasing the "inflight" token, which should prevent this impacting the critical path in the apiserver.

I don't know the http stack well enough to know if this would still block request completion. It seems bad to contend on a single write lock in the handler path for all requests

Reply:
Will moving this to separate go routine fix that?

Reply:
> that would only happen on requests that bumped the high water mark... the majority would be a single atomic read.

missed that this gets reset when the metric is reported. yeah... this approach wouldn't be great.